### ISSUE TITLE
Events data

### ISSUE THREAD

**Issue by gottfrois** [arkency/rails_event_store#2]
Hey guys,

What are your thoughts regarding how event classes are defined? Today we put our event data under a `data` key. Unfortunately we then have to define some helper methods to access this data as method calls instead of hash attributes:

```ruby
class SomeEvent < RailsEventStore::Event  
  def uid
    data.fetch(:uid)
  end

  def foo
    data.fetch(:foo)
  end

  def self.create(uid, foo)
    new(data: { uid: uid, foo: foo })
  end
end
```

Maybe we could drop the `data` namespace and define our events using activemodel:

```ruby
class SomeEvent < RailsEventStore::Event  
  include ActiveModel::Model

  attr_accessor :uid, :foo

  def self.create(uid, foo)
    new(uid: uid, foo: foo)
  end
end
```

The `include` could then be part of `RailsEventStore::Event` class. The downside is with attributes conflicts is the one defined by you (metadata) but this seems a small issue compare to what we can win here. Also we could use `__` methods prefix or similar for event's special attributes.

We can push this even further and define the following method in the parent class to have back a `data` like method we can use in our projections:

```ruby
def data
  attributes_except :event_id, :event_type, :metadata
end

def attributes_except(*attrs)
  as_json.reject { |k, _| attrs.include?(k) }
end
```

```ruby
# from a projection
def fooed(event)
  MyModel.find_by_uid(event.uid).update!(event.data)
end
```

Would love your thoughts on this :)

**Comment by mpraglowski** [comment]
Interesting idea - let me play with it over the weekend and I will post my thought here.

**Comment by gottfrois** [comment]
Did you have a chance to play with this?

**Comment by mpraglowski** [comment]
Not finished yet. I like the idea but I'm concerned about adding additional code to RES gem.

**Comment by gottfrois** [comment]
From what I see, it will depend from activemodel which we already have with rails. I just find the way events are defined today, not very DRY :)

**Comment by mpraglowski** [comment]
Show me the code! (what should be translated here: PR please) :)

**Comment by mpraglowski** [comment]
Resolved in #16
